src/Std/Time/Zoned/ZonedDateTime.lean

```diff
@@ -0,0 +1,280 @@
+  Sigma.mk dt.fst (dt.snd.subMonthsRollOver months)
+
+/--
+Add `Year.Offset` to a `ZonedDateTime`, this function rolls over any excess days into the following
+month.
+-/
+@[inline]
+def addYearsRollOver (dt : ZonedDateTime) (years : Year.Offset) : ZonedDateTime :=
+  Sigma.mk dt.fst (dt.snd.addYearsRollOver years)
+
+/--
+Add `Year.Offset` to a `ZonedDateTime`, it clips the day to the last valid day of that month.
+-/
+@[inline]
+def addYearsClip (dt : ZonedDateTime) (years : Year.Offset) : ZonedDateTime :=
+  Sigma.mk dt.fst (dt.snd.addYearsClip years)
+
+/--
+Add `Hour.Offset` to a `ZonedDateTime`, adjusting the date if necessary.
+-/
+def addHours (dt : ZonedDateTime) (hours : Hour.Offset) : ZonedDateTime :=
+  Sigma.mk dt.fst (dt.snd.addHours hours)
+
+/--
+Subtract `Hour.Offset` from a `ZonedDateTime`, adjusting the date if necessary.
+-/
+@[inline]
+def subHours (dt : ZonedDateTime) (hours : Hour.Offset) : ZonedDateTime :=
+  Sigma.mk dt.fst (dt.snd.subHours hours)
+
+/--
+Add `Minute.Offset` to a `ZonedDateTime`, adjusting the date if necessary.
+-/
+def addMinutes (dt : ZonedDateTime) (minutes : Minute.Offset) : ZonedDateTime :=
+  Sigma.mk dt.fst (dt.snd.addMinutes minutes)
+
+/--
+Subtract `Minute.Offset` from a `ZonedDateTime`, adjusting the date if necessary.
+-/
+@[inline]
+def subMinutes (dt : ZonedDateTime) (minutes : Minute.Offset) : ZonedDateTime :=
+  Sigma.mk dt.fst (dt.snd.subMinutes minutes)
+
+/--
+Add `Second.Offset` to a `ZonedDateTime`, adjusting the date if necessary.
+-/
+def addSeconds (dt : ZonedDateTime) (seconds : Second.Offset) : ZonedDateTime :=
+  Sigma.mk dt.fst (dt.snd.addSeconds seconds)
+
+/--
+Subtract `Second.Offset` from a `ZonedDateTime`, adjusting the date if necessary.
+-/
+@[inline]
+def subSeconds (dt : ZonedDateTime) (seconds : Second.Offset) : ZonedDateTime :=
+  Sigma.mk dt.fst (dt.snd.subSeconds seconds)
+
+/--
+Determines the era of the given `PlainDate` based on its year.
+-/
+@[inline]
+def era (date : ZonedDateTime) : Year.Era :=
+  if date.year.toInt ≥ 0 then
+    .ce
+  else
+    .bce
+
+/--
+Checks if the `DateTime` is in a leap year.
+-/
+def inLeapYear (date : ZonedDateTime) : Bool :=
+  date.year.isLeap
+
+instance : ToTimestamp ZonedDateTime where
+  toTimestamp dt := dt.toTimestamp
+
+instance : HAdd ZonedDateTime (Day.Offset) ZonedDateTime where
+  hAdd := addDays
+
+instance : HSub ZonedDateTime (Day.Offset) ZonedDateTime where
+  hSub := subDays
+
+instance : HAdd ZonedDateTime (Hour.Offset) ZonedDateTime where
+  hAdd := addHours
+
+instance : HSub ZonedDateTime (Hour.Offset) ZonedDateTime where
+  hSub := subHours
+
+instance : HAdd ZonedDateTime (Minute.Offset) ZonedDateTime where
+  hAdd := addMinutes
+
+instance : HSub ZonedDateTime (Minute.Offset) ZonedDateTime where
+  hSub := subMinutes
+
+instance : HAdd ZonedDateTime (Second.Offset) ZonedDateTime where
+  hAdd := addSeconds
+
+instance : HSub ZonedDateTime (Second.Offset) ZonedDateTime where
+  hSub := subSeconds
+
```

Comment:
Should there be `HSub ZonedDateTime ZonedDateTime Duration`?

Reply:
Added instances for `ZonedDateTime`, `DateTime`, `PlainDateTime` and `PlainDate`.